packages/superset-ui-core/src/models/Registry.ts

```diff
@@ -11,19 +11,30 @@ interface ItemWithValue<T> {
 }
 
 interface ItemWithLoader<T> {
-  loader: () => T | Promise<T>;
+  loader: () => T;
 }
 
 export interface RegistryConfig {
   name?: string;
   overwritePolicy?: OverwritePolicy;
 }
 
-export default class Registry<V> {
+/**
+ * Registry class
+ *
+ * Can use generic to specify type of item in the registry
+ * @type V Type of value
+ * @type W Type of value returned from loader function when using registerLoader().
+ * W can be either V, Promise<V> or V | Promise<V>
+ * Set W=V when does not support asynchronous loader.
+ * By default W is set to V | Promise<V> to support
+ * both synchronous and asynchronous loaders.
+ */
+export default class Registry<V, W extends V | Promise<V> = V | Promise<V>> {
```

Comment:
I agree with the added complexity. The high-level idea is to be able to define synchronous registry. I am considering adding a subclass of `Registry` just to alias and hide this multiple generics. 

```
class SyncRegistry<V> extends Registry<V, V>{}
```

`W` affect the `.get()` function in particular if the returned value is always the value or can also be a promise of the value. If it can be either promise of value then the subclass has to always consider if the value is promise or not. Knowing that the returned type is not a promise simplifies that.